lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,394 @@
+
+				if ( ! empty( $s['class_names'] ) ) {
+					$classes = $tags->get_attribute( 'class' );
+					if ( null === $classes ) {
+						continue;
+					}
+
+					foreach ( $s['class_names'] as $class_name ) {
+						if ( ! preg_match( "~\b{$class_name}\b~", $classes ) ) {
+							continue 2;
+						}
+					}
+				}
+
+				if ( isset( $s['hash'] ) && $s['identifier'] !== $tags->get_attribute( 'id' ) ) {
+					continue;
+				}
+
+				if ( isset( $s['has_attribute'] ) && null === $tags->get_attribute( $s['has_attribute'] ) ) {
+					continue;
+				}
+
+				return $tags;
+			}
+		}
+
+		return null;
+	}
+
+	public static function parse_definition( $definition ) {
+		if ( empty( $definition['source'] ) ) {
+			return 'not-sourced';
+		}
+
+		$source = $definition['source'];
+		if ( 'html' !== $source && 'attribute' !== $source ) {
+			return 'unsupported';
+		}
+
+		if ( 'attribute' === $source && empty( $definition['selector'] ) ) {
+			return null;
+		}
+
+		if ( 'html' === $source && empty( $definition['selector'] ) ) {
+			return 'inner-html';
+		}
+
+		$selectors = self::parse_full_selector( $definition['selector'] );
+		if ( null === $selectors ) {
+			return 'unsupported';
+		}
+
+		if ( 'html' === $source ) {
+			return array( 'type' => 'html', 'selector' => $selectors );
+		}
+
+		$attribute = self::parse_attribute( $definition['attribute'] );
+		if ( null === $attribute ) {
+			return null;
+		}
+
+		return array( 'type' => 'attribute', 'selector' => $selectors, 'attribute' => $attribute );
+	}
+
+	public static function parse_full_selector( $s ) {
+		$selectors = [];
+		$at = 0;
+
+		while ( $at < strlen( $s ) ) {
+			$at += strspn( $s, " \f\n\r\t", $at );
+
+			list( $selector, $next_at ) = self::parse_selector( $s, $at );
+			if ( null === $selector ) {
+				return null;
+			}
+
+			$selectors[] = $selector;
+			$at = $next_at;
+
+			if ( $at < strlen( $s ) && ',' !== $s[ $at ] ) {
+				return null;
+			}
+			$at++;
+		}
+
+		return $selectors;
+	}
+
+	public static function parse_selector( $s, $at = 0, $selector = [] ) {
+		$is_first = true;
+
+		while ( $at < strlen( $s ) && ',' !== $s[ $at ] ) {
+			/*
+			 * Descendant combinators are harder to discover because we
+			 * always have to skip whitespace, but that whitespace could
+			 * be the combinator if we don't approach anything else first.
+			 */
+			$ws_length = strspn( $s, " \f\n\r\t", $at );
+			$at += $ws_length;
+
+			if ( !$is_first && $ws_length > 0 && 0 === strspn( $s[ $at ], '>+~' ) ) {
+				$at--;
+				$s[ $at ] = ' ';
+			}
+			$is_first = false;
+
+			switch ( $s[ $at ] ) {
+				case '>':
+				case '+':
+				case '~':
+				case ' ':
+					$combinator = $s[ $at ];
+					$at++;
+					$at += strspn( $s, " \f\n\r\t", $at );
+					$inner = self::parse_selector( $s, $at );
+					if ( null === $inner ) {
+						return null;
+					}
+					list( $inner_selector, $next_at ) = $inner;
+					$inner_selector['combinator'] = $combinator;
+					$selector['then'] = $inner_selector;
```

Comment:
Is the output like a linked list of selectors?

Reply:
that's right. we need to represent selectors in some way that captures the recursive or monadic behavior of the combinators. we have to use the first segment, or _simple selector sequence_, to find a given location in the document, and then _from that starting point_ and in whatever relation to it is appropriate given the combinator, find the next one.

we could probably do this as a list, but I already had set the top level as a list to represent the comma-separated `selectors_group`, which is special because the OR operation here only exists at the top level.

recursing seemed easiest since I could presumably pass a single object to the query generator and it would contain the next simple selector sequence in the chain, or nothing at all.

haven't built that part of the query generator yet though.

Reply:
Not yet sold on supporting `:not()` at this stage.
The list works, but the nesting is convenient because we don't have to track where in the list we are.

What do you see as the benefit of the flat list?